timemachine/md/smc.py

```diff
@@ -0,0 +1,179 @@
+
+    # main loop
+    for (lam_initial, lam_target) in zip(lambdas[:-2], lambdas[1:-1]):
+        # update log weights
+        incremental_log_weights = log_prob(sample_traj[-1], lam_target) - log_prob(sample_traj[-1], lam_initial)
+        log_weights += incremental_log_weights
+
+        # resample
+        indices, log_weights = resample(log_weights)
+        resampled = [sample_traj[-1][i] for i in indices]
+
+        # propagate
+        samples = propagate(resampled, lam_target)
+
+        # log
+        accumulate_results(samples, indices, log_weights, incremental_log_weights)
+
+    # final result: a collection of samples, with associated log weights
+    incremental_log_weights = log_prob(samples, lambdas[-1]) - log_prob(samples, lambdas[-2])
+    incremental_log_weights_traj.append(incremental_log_weights)
+    log_weights_traj.append(np.array(log_weights + incremental_log_weights))
+
+    # cast everything (except samples list) to arrays
+    trajs_dict = dict(
+        traj=sample_traj,
+        log_weights_traj=np.array(log_weights_traj),
+        ancestry_traj=np.array(ancestry_traj),
+        incremental_log_weights_traj=np.array(incremental_log_weights_traj),
+    )
+    return trajs_dict
+
+
+def null_resample(log_weights):
+    """No interaction"""
+    indices = np.arange(len(log_weights))
+    return indices, log_weights
+
+
+def multinomial_resample(log_weights):
+    """sample proportional to exp(log_weights), with replacement"""
+    weights = np.exp(log_weights - logsumexp(log_weights))
+    n = len(log_weights)
+    indices = np.random.choice(np.arange(n), size=n, p=weights)
+
+    # new weights
+    avg_log_weights = logsumexp(log_weights - np.log(n)) * np.ones(n)
+
+    return indices, avg_log_weights
+
+
+def effective_sample_size(log_weights):
+    r"""Effective sample size, taking values in interval [1, len(log_weights)]
+
+    Notes
+    -----
+    * This uses the conventional definition ESS(w) = 1 / \sum_i w_i^2, which has some important known limitations!
+    * See [Elvira, Martino, Robert, 2018] "Rethinking the effective sample size" https://arxiv.org/abs/1809.04129
+        and references therein for some insightful discussion of limitations and possible improvements
+    """
+    weights = np.exp(log_weights - logsumexp(log_weights))
+    return 1 / np.sum(weights ** 2)
+
+
+def fractional_effective_sample_size(log_weights):
+    """effective sample size, normalized to interval (0, 1]"""
+    n = len(log_weights)
+    return effective_sample_size(log_weights) / n
+
+
+def conditional_multinomial_resample(log_weights, thresh=0.5):
+    """if fractional_effective_sample_size(log_weights) < thresh, then multinomial_resample"""
+    if fractional_effective_sample_size(log_weights) < thresh:
+        return multinomial_resample(log_weights)
+    else:
+        return null_resample(log_weights)
+
+
+def refine_samples(samples, log_weights, propagate, lam):
```

Comment:
nit: annotation for `propagate` seems useful here (and in general for arguments of function type)

Reply:
Agreed -- annotations added in https://github.com/proteneer/timemachine/pull/718/commits/5baafa71d9e482035fc7388da3d6063d741615fd